### DISCUSSION TITLE
Having issues passing environment variables to Laravel

### DISCUSSION THREAD

**Discussion by alexjustesen** [serversideup/docker-php#82] [Q&A] [answered]
Hey guys!

I'm having an issue passing environment vars into the container which is causing migrations to fail. I'm building a new image based on `beta-8.1-fpm-nginx` currently and adding a few additional dependencies.

```dockerfile
FROM serversideup/php:beta-8.1-fpm-nginx

ENV PHP_POOL_NAME=speedtest-tracker_php
ENV PHP_POST_MAX_SIZE=1G
ENV PHP_UPLOAD_MAX_FILE_SIZE=1G

# Install addition packages
RUN apt-get update && apt-get install -y \
    cron \
    gnupg \
    php8.1-pgsql \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* /usr/share/doc/*

# Install Speedtest cli
RUN curl -s https://packagecloud.io/install/repositories/ookla/speedtest-cli/script.deb.sh | bash \
    && apt-get install -y speedtest

# Copy package configs
COPY --chmod=644 docker/deploy/cron/scheduler /etc/cron.d/scheduler
COPY --chmod=755 docker/deploy/etc/s6-overlay/ /etc/s6-overlay/

# Copy app
COPY . /var/www/html
COPY .env.docker .env

# Install app dependencies
RUN composer install --no-interaction --prefer-dist --optimize-autoloader --no-dev \
    && mkdir -p /app \
    && mkdir -p storage/logs \
    && php artisan optimize:clear \
    && chown -R webuser:webgroup /var/www/html \
    && crontab /etc/cron.d/scheduler
```

During this build I'm copying a Laravel `.env.docker` file to `.env` so Laravel picks this up and then while running the container passing vars into it.
```bash
docker run -d --name speedtest-tracker \
    -e "DB_CONNECTION=mysql" \
    -e "DB_HOST=host.docker.internal" \
    -e "DB_PORT=3306" \
    -e "DB_DATABASE=speedtest_tracker" \
    -e "DB_USERNAME=root" \
    -e "DB_PASSWORD=password" \
    -v /path/to/app/_data:/app \
    speedtest-tracker
```

Laravel doesn't seem to be seeing any of the vars I'm passing into the container, any help would be greatly appreciated! Reference issue and full code: https://github.com/alexjustesen/speedtest-tracker/pull/41

**Comment by jaydrogers** [comment]
I spent my morning diving in deep to see if I could replicate this. Long story short, I unfortunately could not get it to replicate.

### What I did 
- [x] Created brand new Laravel Install
- [x] Confirmed ENVs are accessible via `docker compose`
- [x] Confirmed ENVs are accessible via `docker run`

### 🎥 See the video demoing this
https://www.dropbox.com/s/t2oucxf6mxxxi9r/LaravelEnvTest.mp4?dl=0

### My Source code
All accessible here: https://github.com/jaydrogers/laravel-env-test

### My gut feeling
Without diving into your app too deep, the only thing that threw me off guard was this:

```Dockerfile
# Copy app
COPY .env.docker .env
```

Theoretically that should work, but I was not able to replicate the issue in a blank laravel install.

Take a look at this and let me know your thoughts and if you have any more questions!

**Reply by alexjustesen** [reply]
Thanks for putting the time into this Jay! So here's the kicker, I can run `php artisan migrate` no issue when the container is running but during the `startup` script it's like it's not aware of any environment variables?

![image](https://user-images.githubusercontent.com/1144087/194162062-b10e3f23-ac09-4d9e-a843-bc014c22f745.png)

**Comment by jaydrogers** [comment, marked as answer]
Whoa dude, this was SUPER HELPFUL!

All that effort in isolating it to a brand new Laravel install paid off troubleshooting this.

### How I was able to replicate
Once I added the auto migrations, then I was able to replicate the issue: https://github.com/jaydrogers/laravel-env-test/commit/4438f2deacd0aa11df784745e38b10cd1ac1978c?short_path=b335630#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

### The problem
In my script, I had a `su - webuser` in my"laravel-automations" script, which was running commands as `webuser` but not passing the environment variables from the container 🙃

### The Solution
I just pushed a commit (https://github.com/serversideup/docker-php/commit/b0716db934fc4d774e297dbdafcd8cb6e1df6f03) that utilizes the `s6-setuidgid` command, which properly passes the environment and switches users "The S6 Way".

### Next Steps
Try again with the latest beta image, and this should work for you 👍

Thanks for testing out my new images 🙌

**Reply by alexjustesen** [reply]
Glad I could "help"! Thanks a lot for taking the time looking into this.
